Update readme (#94)
datastax/kaap: README.md

Some of the key features and benefits of the KAAP operator include:

- **Easy Deployment**: Deploying an Apache Pulsar cluster on Kubernetes is simplified through declarative configurations and automation provided by the operator.

- **Scalability**: The operator enables effortless scaling of Pulsar clusters by automatically handling the creation and configuration of new Pulsar brokers and bookies as per defined rules.

- **High Availability**: The operator implements best practices for high availability, ensuring that Pulsar clusters are fault-tolerant and can sustain failures without service disruptions.

- **Monitoring and Metrics**: Integrated monitoring and metrics collection allow users to gain insights into the health and performance of Pulsar clusters, making it easier to detect and resolve issues.

- **Lifecycle Management**: The operator takes care of common Pulsar cluster lifecycle tasks, such as cluster creation, upgrade, configuration updates, and graceful shutdowns.

> - **Integration with Kubernetes Ecosystem**: The Apache Pulsar Kubernetes Operator seamlessly integrates with other Kubernetes-native tools and features, such as Kubernetes RBAC, Prometheus, and Grafana, enabling a cohesive and comprehensive ecosystem for managing Pulsar clusters.

Reviewer: ```suggestion
- **Integration with Kubernetes Ecosystem**: it seamlessly integrates with other Kubernetes-native tools and features, such as Kubernetes RBAC, Prometheus, and Grafana, enabling a cohesive and comprehensive ecosystem for managing Pulsar clusters.
```

Whether you are a developer looking to leverage the power of Apache Pulsar in your Kubernetes environment or an operator seeking to streamline the management of Pulsar clusters, the KAAP Operator provides a robust and user-friendly solution. In this documentation, we will explore the installation, configuration, and usage of the operator, empowering you to harness the full potential of Apache Pulsar within your Kubernetes deployments.

## Documentation

Full documentation is available in the [DataStax Streaming Documentation](https://docs.datastax.com/en/streaming/kaap-operator/index.html) or at [this repo's GitHub Pages site](https://datastax.github.io/kaap/docs/).

> ## Getting started

Reviewer: I'd add an explanation how to get the Pulsar cluster ready, not only the operator

Reviewer 2: I agree. Can we add a quick start example for creating a pulsarcluster and making sure its running?

Reviewer: ```
helm repo add kaap https://datastax.github.io/kaap
helm repo update
curl -LOs https://raw.githubusercontent.com/datastax/kaap/main/helm/examples/dev-cluster/values.yaml
helm install pulsar -f values.yaml 
kubectl wait pulsar pulsar --for condition=Ready=True --timeout=240s
```

1. Install the DataStax Helm repository:
```
helm repo add kaap https://datastax.github.io/kaap
helm repo update
```
2. Install the Pulsar operator Helm chart:
```
helm install kaap kaap/kaap
```
Result:
```
NAME: kaap
LAST DEPLOYED: Wed Jun 28 11:37:45 2023
NAMESPACE: pulsar-cluster
STATUS: deployed
REVISION: 1
TEST SUITE: None
```

> 3. Ensure pulsar-operator is up and running:

Reviewer: ```suggestion
3. Ensure the operator is up and running:
```

```
kubectl get deployment
```
Result:
```
NAME              READY   UP-TO-DATE   AVAILABLE   AGE
kaap   1/1     1            1           13m
```

4. You've now installed KAAP.
By default, when KAAP is installed, the PulsarCluster CRDs are also created.

> This setting is defined in the Pulsar operator values.yaml file as `crd: create: true`.

Reviewer: ```suggestion
This setting is defined in the KAAP values.yaml file as `crd: {create: true}`.
```

5. To see available CRDs:
```

> kubectl get crds | grep datastax

Reviewer: ```suggestion
kubectl get crds | grep kaap
```

```
Result:
```

> autorecoveries.pulsar.oss.datastax.com           2023-05-12T16:35:59Z

Reviewer: all of them are now with kaap instead of pulsar

autorecoveries.kaap.oss.datastax.com

bastions.pulsar.oss.datastax.com                 2023-05-12T16:36:00Z
bookkeepers.pulsar.oss.datastax.com              2023-05-12T16:36:00Z
brokers.pulsar.oss.datastax.com                  2023-05-12T16:36:01Z
functionsworkers.pulsar.oss.datastax.com         2023-05-12T16:36:01Z
proxies.pulsar.oss.datastax.com                  2023-05-12T16:36:02Z
pulsarclusters.pulsar.oss.datastax.com           2023-05-12T16:36:04Z
zookeepers.pulsar.oss.datastax.com               2023-05-12T16:36:06Z
```

## Uninstall KAAP

To uninstall KAAP:
```
helm delete kaap
```
Result:
```
release "kaap" uninstalled
```

For more, see the [DataStax Streaming Documentation](https://docs.datastax.com/en/streaming/kaap-operator/index.html) or [this repo's GitHub Pages site](https://datastax.github.io/kaap/docs/).

> ## Build KAAP from source

Reviewer: this is not useful for users, I'd drop it

# Kubernetes Autoscaling for Apache Pulsar (KAAP)

> Kubernetes Autoscaling for Apache Pulsar (KAAP) simplifies running [Apache Pulsar](https://pulsar.apache.org) on Kubernetes by applying the familiar [Operator pattern](https://kubernetes.io/docs/concepts/extend-kubernetes/operator/) to Pulsar's components.

Reviewer 2: Can we add more about the autoscaling capabilities that KAAP provides? This is an innovative feature in this space, and I think we want to trumpet it. 

Specifically, I'd like to see some description about how KAAP will automatically scale brokers up and down based on resource usage like CPU and memory. Broker autoscaling is integrated with the Pulsar broker load balancer which has insight into load on all brokers so it makes smarter decisions about when to add remove brokers, as opposed to simple solutions based on Kubernetes HorizontalPodAutoscaler.

KAAP also can automatically scale up BookKeeper nodes in response to running low on storage and because of BookKeepers segment-based design that storage is immediately available for use by the cluster. If storage usage it low, it will automatically scale down BookKeeper nodes to free up volume usage in cloud environments, saving storage costs. KAAP does this scale down in controlled, safe manner ensuring no data loss and guaranteeing the configuration replication factor for all message. For example if using a replication factor of 3 (write and ack quorum of 3), at all times during the scale down 3 replicas are maintained to ensure data can be recovered even if there is a failure during the scale down phase.

Can we add some text along these lines?

- **Monitoring and Metrics**: Integrated monitoring and metrics collection allow users to gain insights into the health and performance of Pulsar clusters, making it easier to detect and resolve issues.

- **Lifecycle Management**: The operator takes care of common Pulsar cluster lifecycle tasks, such as cluster creation, upgrade, configuration updates, and graceful shutdowns.

- **Integration with Kubernetes Ecosystem**: it seamlessly integrates with other Kubernetes-native tools and features, such as Kubernetes RBAC, Prometheus, and Grafana, enabling a cohesive and comprehensive ecosystem for managing Pulsar clusters.

Whether you are a developer looking to leverage the power of Apache Pulsar in your Kubernetes environment or an operator seeking to streamline the management of Pulsar clusters, the KAAP Operator provides a robust and user-friendly solution. In this documentation, we will explore the installation, configuration, and usage of the operator, empowering you to harness the full potential of Apache Pulsar within your Kubernetes deployments.

> 

Reviewer 2: I think we should explain the relationship between this operator and the helm chart that installs the whole production stack (Grafana, Prometheus, etc, etc). We should link to that Helm chart.

Reviewer 2: I see now that the Helm chart is part of this repo. Given that, I think we need to be clear in this readme that it contains both the operator and KAAP stack. We should have a section on the KAAP stack and some quick start information for that.
